### PR TITLE
chore: move artifacts recording to TestLifecycleInstrumentation

### LINKED ISSUE #30287: [Bug]: trace not recorded for "workerStorageState"
### Version

18.2.3

### Steps to reproduce

1. clone my repo - https://github.com/Danieliverant/playwright-no-trace
2. `cd playwright-no-trace`
3. `nx e2e e2e --trace on`
4. `npx playwright show-trace ./e2e/playwright-report/example-has-title-chromium/trace.zip`

(the test will fail)

### Expected behavior

To see the steps in the "wrokerStorageState"

### Actual behavior

Not seeing anything from the "wrokerStorageState"

image

### Additional context

Same thing for UI mode - you wait until the "workerStorageState" to finish and only then you see the trace.

### Environment

```shell
System:
    OS: macOS 13.6
    CPU: (10) arm64 Apple M1 Pro
    Memory: 85.05 MB / 16.00 GB
  Binaries:
    Node: 18.10.0 - ~/.nvm/versions/node/v18.10.0/bin/node
    npm: 8.19.2 - ~/.nvm/versions/node/v18.10.0/bin/npm
  Languages:
    Bash: 3.2.57 - /bin/bash
  npmPackages:
    @playwright/test: ^1.36.0 => 1.43.0
```

### LINKED ISSUE #30718: [Bug]: Timeline & preview disappear after timeout in fixture cleanup
### Version

1.44.0

### Steps to reproduce

Edit: Minimal reproduction: https://github.com/tobz1000/playwright-30718

1. Create a test fixture which causes an error during the cleanup phase. Edit: The fixture should have a timeout configured, and the error must cause faillure by timeout.
2. Create a test which uses the fixture.
3. Run the test UI and start the test (I reproduced via UI port): `npx playwright test --ui-port 3031`
4. Once the test has failed during fixture cleanup, the timeline and preview window will disappear:
image

~~I can create a minimal reproduction repo if requested.~~

### Expected behavior

Timeline and window preview should be visible once the test has finished.

### Actual behavior

Timeline and window preview are not visible.

### Additional context

A similar bug was reported and puportedly fixed in a previous version: #29826

### Environment

```shell
System:
  OS: Linux 5.15 Debian GNU/Linux 11 (bullseye) 11 (bullseye)
  CPU: (12) x64 AMD Ryzen 5 3600 6-Core Processor
  Memory: 6.89 GB / 15.63 GB
  Container: Yes
Binaries:
  Node: 20.12.2 - /usr/local/share/nvm/versions/node/v20.12.2/bin/node
  Yarn: 1.22.19 - /usr/bin/yarn
  npm: 10.5.0 - /usr/local/share/nvm/versions/node/v20.12.2/bin/npm
  pnpm: 9.0.3 - /usr/local/share/nvm/versions/node/v20.12.2/bin/pnpm
  bun: 1.1.4 - ~/.bun/bin/bun
IDEs:
  VSCode: 1.88.1 - /vscode/vscode-server/bin/linux-x64/e170252f762678dec6ca2cc69aba1570769a5d39/bin/remote-cli/code
Languages:
  Bash: 5.1.4 - /bin/bash
```

### LINKED ISSUE #30959: [Question]: Screenshot is missing, when Authenticating in UI mode failed using fixtures
### Context
I copied the [fixture](https://playwright.dev/docs/auth) for Authenticating from the docs, and I added a failing step. I see a missing screenshot in the HTML report
Also, I found this issue https://github.com/microsoft/playwright/issues/14544, but I think we need to find any way to display screenshots or video

### Version

latest

### Steps to reproduce

1. Clone my repo https://github.com/AlexeyCL/playwright.git
2. npm install
3. npx playwright test 

### Expected behavior

I should see screenshota 

### Actual behavior

 In HTML report screenshot and video are missing 

### Additional context

I found this issue https://github.com/microsoft/playwright/issues/14544, but I think we need to find any way to display screenshot or video

### Environment

```shell
System:
All
```

### PR DESCRIPTION
**Pull request by dgozman**
The spirit of this change is reverting #23153. Since that time, we have moved tracing and `artifactsDir` lifetime into the test runner, so the reason for revert is mitigated.

Fixes #30287, fixes #30718, fixes #30959.

### CODE REVIEW

File: packages/playwright/src/worker/workerMain.ts [hunk 1]
```diff
@@ -372,10 +374,10 @@ export class WorkerMain extends ProcessRunner {
 
       try {
         // Run "immediately upon test function finish" callback.
-        await testInfo._runAsStage({ title: 'on-test-function-finish', runnable: { type: 'test', slot: afterHooksSlot } }, async () => testInfo._onDidFinishTestFunction?.());
+        await testInfo._runAsStage({ title: 'on-test-function-finish', runnable: { type: 'test', slot: tracingSlot } }, async () => {
+          await testLifecycleInstrumentation()?.onTestFunctionEnd?.();
+        });
       } catch (error) {
-        if (error instanceof TimeoutManagerError)
-          didTimeoutInAfterHooks = true;
```
Comment (yury-s) [inline packages/playwright/src/worker/workerMain.ts:378]: is this related to the change?

File: packages/playwright/src/index.ts [hunk 2]
```diff
@@ -45,23 +44,99 @@ if ((process as any)['__pw_initiator__']) {
...
+      },
+      onWillCloseRequestContext: async (context: APIRequestContext) => {
+        await artifactsRecorder?.willCloseRequestContext(context);
+      },
+    };
+
+    const clientInstrumentation = (_playwrightImpl as any)._instrumentation as ClientInstrumentation;
+    clientInstrumentation.addListener(csiListener);
+
+    if (currentTestInfo()) {
```
Comment (yury-s) [inline packages/playwright/src/index.ts:123]: currentTestInfo() should always be defined here, remove the if?

File: packages/playwright/src/index.ts [hunk 3]
```diff
@@ -45,23 +44,99 @@ if ((process as any)['__pw_initiator__']) {
...
+      // Setup of this fixture happens after the first test starts.
+      await instrumentation.onTestBegin?.();
+    }
+
+    await use(_playwrightImpl);
+
+    if (currentTestInfo()) {
+      // Teardown of this fixture might happen right before the failing test ends
+      // as a part of worker cleanup.
```
Comment (yury-s) [inline packages/playwright/src/index.ts:132]: I'd rephrase it to something along the lines: "If the test fails, worker teardown will destroy this fixture before `onTestEnd();` is called, so we call it here"

**Review by yury-s — APPROVED** [review]

File: packages/playwright/src/worker/workerMain.ts (hunk 1 above)
Comment (dgozman) [reply packages/playwright/src/worker/workerMain.ts:378]: Yes, this is because we now pass a different `tracingSlot`, so timing out there does not prevent after hooks from running.

**Review by yury-s — APPROVED** [review]
